src/generator/qbe.rs

```diff
@@ -0,0 +1,1221 @@
+        // TODO: validate the same in non-void ones
+        if func.ret_type.is_none() {
+            qfunc.add_instr(QbeInstr::Ret(None));
+        }
+
+        self.scopes.pop();
+
+        Ok(qfunc)
+    }
+
+    /// Generates a statement
+    fn generate_statement(
+        &mut self,
+        func: &mut QbeFunction,
+        stmt: &Statement,
+    ) -> GeneratorResult<()> {
+        match stmt {
+            Statement::Block(statements, _) => {
+                self.scopes.push(HashMap::new());
+                for stmt in statements.iter() {
+                    self.generate_statement(func, stmt)?;
+                }
+                self.scopes.pop();
+            }
+            Statement::Declare(var, expr) => {
+                let ty = self.get_type(
+                    var.ty
+                        .as_ref()
+                        .ok_or_else(|| format!("Missing type for variable '{}'", &var.name))?
+                        .to_owned(),
+                )?;
+                let tmp = self.new_var(&ty, &var.name)?;
+
+                if let Some(expr) = expr {
+                    let (ty, result) = self.generate_expression(func, expr)?;
+                    func.assign_instr(tmp, ty, QbeInstr::Copy(result));
+                }
+            }
+            Statement::Assign(lhs, rhs) => {
+                let (_, rhs) = self.generate_expression(func, rhs)?;
+                // TODO: type check
+                self.generate_assignment(func, lhs, rhs)?;
+            }
+            Statement::Return(val) => match val {
+                Some(expr) => {
+                    let (_, result) = self.generate_expression(func, expr)?;
+                    // TODO: Cast to function return type
+                    func.add_instr(QbeInstr::Ret(Some(result)));
+                }
+                None => func.add_instr(QbeInstr::Ret(None)),
+            },
+            Statement::If(cond, if_clause, else_clause) => {
+                self.generate_if(func, cond, if_clause, else_clause)?;
+            }
+            Statement::While(cond, body) => {
+                self.generate_while(func, cond, body)?;
+            }
+            Statement::Break => {
+                if let Some(label) = &self.loop_labels.last() {
+                    func.add_instr(QbeInstr::Jmp(format!("{}.end", label)));
+                } else {
+                    return Err("break used outside of a loop".to_owned());
+                }
+            }
+            Statement::Continue => {
+                if let Some(label) = &self.loop_labels.last() {
+                    func.add_instr(QbeInstr::Jmp(format!("{}.cond", label)));
+                } else {
+                    return Err("continue used outside of a loop".to_owned());
+                }
+            }
+            Statement::Exp(expr) => {
+                self.generate_expression(func, expr)?;
+            }
+            _ => todo!("statement: {:?}", stmt),
```

Comment:
Could you make this exhaustive?

Reply:
It's already exhaustive. When it reaches statement it doesn't know yet, it will print `not yet implemented: statement: <statement name>`, it's better than sticking a separate `todo!()` on each unknown statement

Reply:
But when adding a new statement type in the parser, there's a chance we forget to update this `match`.
[This](https://github.com/antimony-lang/antimony/pull/42/files#diff-d7fb403429bff994dfbfe09c033aaef341e51d88ab055a3a787e99a967783816R287-R289) is more explicit and will throw an error if we miss it.

Reply:
rustc enforces that `match`es are exhaustive, if we forget something, it won't compile

Reply:
Sorry, I think my wording is wrong. Yes, it's exhaustive, but it's not explicit. `_ => todo!()` as a catch-all only makes sense if there are a lot of options. In this case, the 2/3 (?, I don't actually know how many there are left) enum fields can/should be written out.

Reply:
Right now we have 2 unimplemented statements (`for` and `match`) and 3 unimplemented expressions (array initialization, array access and `self`). I think it's fine to keep a catch-all for now just to not litter the generator

Reply:
Still, it's good to know that these 5 paths are not implemented yet. And if we add more statements/expressions in the future, we will have to actively hunt down this TODO, whereas if it's explicit, it won't even compile and the developer cannot forget about it.

Reply:
You don't need to hunt it down in the current state either, because it says which statement/expression is not implemented

Reply:
Yes, at runtime, even though it could've been prevented at compiletime.

Reply:
I prefer to leave this as-is and continue working on arrays :)